apps/builder/app/builder/features/props-panel/controls/text.tsx

```diff
@@ -1,80 +1,102 @@
 };
 
-const AsInput = ({
-  label,
-  id,
+/**
+ * UniversalInput component allows to dynamically decide if we need a single-line input or a multi-line textarea.
+ * Problem it is solving is that we don't know from data type string if user needs a textarea or an input.
+ *
+ * Single-line mode:
+ * - enter saves
+ * - shift+enter - adds newline and saves
+ *
+ * Multiline-mode:
+ * - enter saves and ads newline
+ * - cmd+enter - saves, no newline
+ */
+const UniversalInput = ({
   value,
+  defaultRows,
+  maxRows = 10,
   onChange,
-  onDelete,
-}: ImplementationProps) => {
-  const localValue = useLocalValue(value, onChange);
-
-  return (
-    <HorizontalLayout label={label} id={id} onDelete={onDelete}>
-      <InputField
-        id={id}
-        value={localValue.value}
-        onChange={(event) => localValue.set(event.target.value)}
-        onBlur={localValue.save}
-        onKeyDown={(event) => {
-          if (event.key === "Enter") {
-            localValue.save();
-          }
-        }}
-        css={{ width: theme.spacing[22] }}
-      />
-    </HorizontalLayout>
+  onSubmit,
+  ...rest
+}: UniversalInputProps) => {
+  const [rows, setRows] = useState<number>(() =>
+    Math.min(defaultRows || countLines(value), maxRows)
   );
-};
 
-const AsTextarea = ({
-  label,
-  id,
-  value,
-  onChange,
-  onDelete,
-  rows,
-}: ImplementationProps & { rows: number }) => {
-  const localValue = useLocalValue(value, onChange);
+  const handleChange = (value: string) => {
+    setRows(Math.min(countLines(value), maxRows));
+    onChange(value);
+  };
 
   return (
-    <VerticalLayout label={label} id={id} onDelete={onDelete}>
-      <Flex css={{ py: theme.spacing[2] }}>
-        <TextArea
-          id={id}
-          value={localValue.value}
-          onChange={(event) => localValue.set(event.target.value)}
-          onBlur={localValue.save}
-          rows={rows ?? 1}
-          onKeyDown={(event) => {
-            if (event.key === "Enter" && event.metaKey) {
-              localValue.save();
-            }
-          }}
-        />
-      </Flex>
-    </VerticalLayout>
+    <TextArea
+      {...rest}
+      css={
+        rows > 1
+          ? { resize: "vertical" }
+          : { resize: "none", whiteSpace: "nowrap", overflow: "hidden" }
+      }
+      defaultValue={value}
+      onChange={(event) => {
+        const { value } = event.target;
+        handleChange(value);
+      }}
+      onBlur={(event) => {
+        handleChange(value);
+        onSubmit();
+      }}
+      rows={rows}
+      onKeyDown={(event) => {
+        if (event.key !== "Enter") {
+          return;
+        }
+        const isMultiline = rows > 1;
+        if (isMultiline === false) {
+          event.preventDefault();
+        }
+        // Insert the newline at the caret position.
+        if (event.shiftKey && isMultiline === false) {
+          const element = event.currentTarget;
+          if (element.selectionStart || element.selectionStart === 0) {
```

Comment:
selectionStart has `unsigned long` type based on spec, so this condition should be always true

Reply:
selectionStart is undefined if nothing is selected

Reply:
just checked, you are right, its always a number